sdk/core/Azure.Core.Experimental/tests/public/JsonDataPublicTests.cs

```diff
@@ -0,0 +1,400 @@
+                Assert.AreEqual(++i, item);
+            }
+        }
+
+        [Test]
+        public void CanGetDynamicFromBinaryData()
+        {
+            var data = new BinaryData(new
+            {
+                array = new[] { 1, 2, 3 }
+            });
+
+            dynamic json = data.ToDynamic();
+            dynamic array = json.array;
+
+            int i = 0;
+            foreach (int item in array)
+            {
+                Assert.AreEqual(++i, item);
+            }
+        }
+
+        [Test]
+        public void EqualsHandlesStringsSpecial()
+        {
+            dynamic json = new BinaryData("\"test\"").ToDynamic();
+
+            Assert.IsTrue(json.Equals("test"));
+            Assert.IsTrue(json.Equals(new BinaryData("\"test\"").ToDynamic()));
+        }
+
+        [Test]
+        public void EqualsForObjectsAndArrays()
+        {
+            dynamic obj1 = new BinaryData(new { foo = "bar" }).ToDynamic();
+            dynamic obj2 = new BinaryData(new { foo = "bar" }).ToDynamic();
+
+            dynamic arr1 = new BinaryData(new[] { "bar" }).ToDynamic();
+            dynamic arr2 = new BinaryData(new[] { "bar" }).ToDynamic();
+
+            // For objects and arrays, Equals provides reference equality.
+            Assert.AreEqual(obj1, obj1);
+            Assert.AreEqual(arr1, arr1);
+
+            Assert.AreNotEqual(obj1, obj2);
+            Assert.AreNotEqual(arr1, arr2);
+        }
+
+        [Test]
+        public void OperatorEqualsForInt()
+        {
+            dynamic fiveJson = new BinaryData("{ \"value\": 5 }").ToDynamic().value;
+            dynamic sixJson = new BinaryData("{ \"value\": 6 }").ToDynamic().value;
+
+            Assert.IsTrue(fiveJson == 5);
+            Assert.IsTrue(5 == fiveJson);
+            Assert.IsFalse(fiveJson != 5);
+            Assert.IsFalse(5 != fiveJson);
+
+            Assert.IsFalse(sixJson == 5);
+            Assert.IsFalse(5 == sixJson);
+            Assert.IsTrue(sixJson != 5);
+            Assert.IsTrue(5 != sixJson);
+        }
+
+        [Test]
+        public void OperatorEqualsForString()
+        {
+            dynamic fooJson = new BinaryData("\"foo\"").ToDynamic();
+            dynamic barJson = new BinaryData("\"bar\"").ToDynamic();
+
+            Assert.IsTrue(fooJson == "foo");
+            Assert.IsTrue("foo" == fooJson);
+            Assert.IsFalse(fooJson != "foo");
+            Assert.IsFalse("foo" != fooJson);
+
+            Assert.IsFalse(barJson == "foo");
+            Assert.IsFalse("foo" == barJson);
+            Assert.IsTrue(barJson != "foo");
+            Assert.IsTrue("foo" != barJson);
+        }
+
+        [Test]
+        public void EqualsForStringNUnit()
+        {
+            dynamic foo = new BinaryData("{ \"value\": \"foo\" }").ToDynamic();
+            var value = foo.Value;
+
+            // TODO: Outstanding question regarding whether we want these to succeed without a cast.
+            Assert.AreEqual(value, "foo");
```

Comment:
This is the same question about whether we should require a cast to a primitive type.  NUnit does not use the provided `==` operator and BindConvert is never invoked.